model/task_queue_service.go

```diff
@@ -33,8 +103,55 @@ type schedulableUnit struct {
 	tasks        []TaskQueueItem
 }
 
-// NewTaskDispatchService creates a taskDispatchService from a slice of TaskQueueItems.
-func NewTaskDispatchService(distroID string, items []TaskQueueItem) *taskDispatchService {
+// NewTaskDispatchService creates a taskDistroDispatchService from a slice of TaskQueueItems.
+func NewDistroTaskDispatchService(distroID string, items []TaskQueueItem, ttl time.Duration) TaskDispatchService {
+	t := &taskDistroDispatchService{
+		distroID: distroID,
+		ttl:      ttl,
+	}
+
+	if items != nil {
+		t.rebuild(items)
+	}
+
+	return t
+}
+
+func (t *taskDistroDispatchService) Refresh() error {
+	if !t.shouldRefresh() {
+		return nil
+	}
+
+	queue, err := FindDistroTaskQueue(t.distroID)
```

Comment:
I think the lock should move out of `rebuild` and above this line. Otherwise the queue found by this line could be stale by the time the lock is acquired in `rebuild`. This also means removing the lock from `shouldRefresh` (which, practically, means removing that function).